tests/integration/test_pip_packages.py

```diff
@@ -24,3 +28,200 @@ def test_failing_pip_local_path(test_env):
+        "pkg_managers": env_data["pkg_managers"],
+    }
+    try:
+        initial_response = client.create_new_request(payload=payload)
+        completed_response = client.wait_for_complete_request(initial_response)
+    finally:
+        # Delete the dependency branch
+        dep_repo_origin.push(refspec=f":{env_data['test_branch']}")
+
+    replace_rules = {
+        "FIRST_DEP_COMMIT": new_dep_commits[0],
+        "SECOND_DEP_COMMIT": new_dep_commits[1],
+        "FIRST_DEP_HASH": dep_hash,
+        "MAIN_REPO_COMMIT": main_repo_commit,
+    }
+    update_expected_data(env_data, replace_rules)
+
+    try:
+        assert_successful_cached_request(completed_response, env_data, tmpdir, client)
+        # Create new Cachito request to test cached deps
+        initial_response = client.create_new_request(payload=payload)
+        completed_response = client.wait_for_complete_request(initial_response)
+    finally:
+        # Delete branch with commit in the main repo
+        main_repo_origin.push(refspec=f":{env_data['test_branch']}")
+    assert_successful_cached_request(completed_response, env_data, tmpdir, client)
+
+
+def assert_successful_cached_request(response, env_data, tmpdir, client):
+    """
+    Provide all verifications for Cachito request with cached dependencies.
+
+    :param Response response: completed Cachito response
+    :param dict env_data: the test data
+    :param tmpdir: the path to directory with testing files
+    :param Client client: the Cachito client to make requests
+    """
+    utils.assert_properly_completed_response(response)
+
+    response_data = response.data
+    expected_response_data = env_data["response_expectations"]
+    utils.assert_elements_from_response(response_data, expected_response_data)
+
+    client.download_and_extract_archive(response.id, tmpdir)
+    source_path = tmpdir.join(f"download_{str(response.id)}")
+    expected_files = env_data["expected_files"]
+    utils.assert_expected_files(source_path, expected_files, tmpdir)
+
+    purl = env_data["purl"]
+    deps_purls = []
+    source_purls = []
+    if "dep_purls" in env_data:
+        deps_purls = [{"purl": x} for x in env_data["dep_purls"]]
+    if "source_purls" in env_data:
+        source_purls = [{"purl": x} for x in env_data["source_purls"]]
+
+    image_contents = [{"dependencies": deps_purls, "purl": purl, "sources": source_purls}]
+    utils.assert_content_manifest(client, response.id, image_contents)
+
+
+def replace_by_rules(orig_str, replace_rules):
+    """
+    Replace elements in string according to replace rules.
+
+    :param str orig_str: original string
+    :param dict replace_rules: replace rules as a dictionary:
+        {<ORIG_PART>: <NEW_PART>}
+    :return: string with replaced values
+    :rtype: str
+    """
+    res_string = copy(orig_str)
+    for s, r in replace_rules.items():
+        if s in res_string:
+            res_string = res_string.replace(s, r)
+    return res_string
```

Comment:
Oh, thank you. I will post an update without `if`. 
But I still need to copy the original string. Otherwise, I need to write something like: 
```suggestion
    res_string = ""
    for s, r in replace_rules.items():
        if len(res_string):
            res_string = res_string.replace(s, r)
        else:
            res_string = orig_string.replace(s, r)
    return res_string
```

Reply:
Or you could just do `res_string = orig_str`, without copying :-)

Reply:
I deleted if condition but kept clone part.

Reply:
Oh, sure, I did not see your last comment.